drivers: working with flash partitions - a MTD API

Laczen opened the pull request:
The storage subsystem provides a flash_map API that allows the user to
work with dts defined partition through a flash_map structure. There is
at present no API to work with dts partitions directly.

The present PR proposes such a direct API.

In order to use a specific partition (e.g. the storage partition) the
following lines are used to create a link to the partition structure:

```
const struct mtd_part *part = MTD_PART_GET(storage);
```

Is a base to solve #34227

Laczen committed:
drivers: mtd api proposal

The PR proposes a API to work with memory technology devices (MTD).

MTD's work with read/write and erase methods. MTD's are supported as
master items (the complete device) and partitions (defined in the dts).
All read/write/erase operation verify that nothing is done outside the
mtd master or partition. Next to the basic read/write/erase functions
some helper functions are also supplied.

MTD partitions can be specified in the dts as read-only. Partitions can
also be sub-partitioned into even smaller partitions.

To work with partitions a mtd_info structure can be requested as
(where storage is a dts defined partition label):
```
const struct mtd_info *mtd = MTD_PARTITION_GET(storage);
```
or as (where flash0 is the nodelabel of a master):
```
const struct mtd_info *mtd = MTD_MASTER_GET(flash0);
```
and a write, read, erase if performed as:
```
rc = mtd_write(const struct mtd_info *mtd, off_t off,
	       const void *data, size_t len);
rc = mtd_read(const struct mtd_info *mtd, off_t off, void *data,
	      size_t len);
rc = mtd_erase(const struct mtd_info *mtd, off_t off, size_t len);
```

To get the `write-block-size` on a mtd item:
```
rc = mtd_get_wbs(const struct mtd_info *mtd, uint8_t *wbs);
```

To get the `erase-block-size` on a mtd item (only available if the
`erase-block-size` is constant over a mtd item):
```
rc = mtd_get_ebs(const struct mtd_info *mtd, size_t *ebs);
```

To get the `erase-default-value` on a mtd_item:
```
rc = mtd_get_edv(const struct mtd_info *mtd, uint8_t *edv);
```

To get the block on a mtd item that a `offset off` belongs to:
```
rc = mtd_get_block_at(const struct mtd_info *part, off_t off,
		      struct mtd_block *block);
```

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
The partition `erase_block_size` is not yet specified in the dts.

Laczen commented:
Added the proposal to dev-review to know if the proposal would be accepted.

de-nordic reviewed (dismissed):
-1 from my perspective: the change duplicates flash map API purpose and  does not evaluate impact on mcuboot, mcumgr and DFU.

Laczen commented:
@de-nordic, I can understand that from your perpective (mcuboot, mcumgr, DFU) this PR gets -1. But maybe your perspective needs some adaptation :-).
The goal of this PR is to provide a simple method for ZEPHYR applications to use partitions. These applications are nvs, fcb, simplefs, ... They all don't need a flashmap description of all partitions that are available, they just need (secured) access to the partition so that there actions are limited to the partition itself.
This PR would reduce the size of applications by:
a. eliminating unused info
b. providing a simple routine for retrieving the erase block size that just requires 1 extra uint32, while removing the code for searching the blocksize and removing the need for CONFIG_FLASH_PAGE. This can just be defined as a partition property in the dts.

Laczen commented:
Added support for read-only partitions

carlescufi commented:
```
Extension error (zephyr.doxyrunner):
Handler <function doxygen_build at 0x7f1d5126bca0> for event 'builder-inited' threw an exception (exception: Doxygen process returned non-zero (1))
FAILED: CMakeFiles/html 
```

This looks like an issue in @gmarull's extension.

gmarull commented:
There's an issue in Doxygen:

```
(.venv) ➜  zephyr git:(flash_partition) ninja -C doc/_build doxygen
ninja: Entering directory `doc/_build'
[1/1] Running /usr/bin/doxygen
FAILED: CMakeFiles/doxygen 
cd /home/gmarull/ws/teslabs/zephyr/zephyr/doc && /usr/bin/doxygen /home/gmarull/ws/teslabs/zephyr/zephyr/doc/_build/zephyr.doxyfile
/home/gmarull/ws/teslabs/zephyr/zephyr/include/storage/flash_partition.h:132: error: argument 'fa' of command @param is not found in the argument list of flash_partition_erase(const struct flash_partition *fp, off_t off, size_t len) (warning treated as error, aborting now)
ninja: build stopped: subcommand failed.
```

I'll check why it's not seen on CI, seems like an issue in the extension.

Laczen commented:
> There's an issue in Doxygen:
> 
> ```
> (.venv) ➜  zephyr git:(flash_partition) ninja -C doc/_build doxygen
> ninja: Entering directory `doc/_build'
> [1/1] Running /usr/bin/doxygen
> FAILED: CMakeFiles/doxygen 
> cd /home/gmarull/ws/teslabs/zephyr/zephyr/doc && /usr/bin/doxygen /home/gmarull/ws/teslabs/zephyr/zephyr/doc/_build/zephyr.doxyfile
> /home/gmarull/ws/teslabs/zephyr/zephyr/include/storage/flash_partition.h:132: error: argument 'fa' of command @param is not found in the argument list of flash_partition_erase(const struct flash_partition *fp, off_t off, size_t len) (warning treated as error, aborting now)
> ninja: build stopped: subcommand failed.
> ```
> 
> I'll check why it's not seen on CI, seems like an issue in the extension.

Thanks, I can fix that.

Laczen commented:
During the dev-review meeting there was a concern by @d3zd3z regarding the definition of a fixed erase block in a partition.

First of all: as the flash device is known in the flash partition there is no limit what so ever to define a `get_sectors` routine.

More important, why making a fixed setting (configurable in devicetree):
a. Most applications use a fixed erase-block-size,
b. WITHIN a partition support of varying block sizes would be highly questionable, this is certainly not supported by any system I have seen.
c. It would allow easy definition of a block-size for use in block-devices.
d. It would allow very easy definition of the number of sectors used by nvs/fcb: size/erase-block-size, no need to do whatever code to step over sectors.
e. By defining it at the partition level there is no need to define it again inside a subsystem (e.g. sector-size and sector-number in nvs). 
f. As the value is valid as long as it is a multiple of the page-size, it is possible to change the value so that it best matches the requirements.

de-nordic commented:
> 
> 
> During the dev-review meeting there was a concern by @d3zd3z regarding the definition of a fixed erase block in a partition.
> 
> First of all: as the flash device is known in the flash partition there is no limit what so ever to define a `get_sectors` routine.
> 
> More important, why making a fixed setting (configurable in devicetree):
> a. Most applications use a fixed erase-block-size,
> b. WITHIN a partition support of varying block sizes would be highly questionable, this is certainly not supported by any system I have seen.
> c. It would allow easy definition of a block-size for use in block-devices.
> d. It would allow very easy definition of the number of sectors used by nvs/fcb: size/erase-block-size, no need to do whatever code to step over sectors.
> e. By defining it at the partition level there is no need to define it again inside a subsystem (e.g. sector-size and sector-number in nvs).
> f. As the value is valid as long as it is a multiple of the page-size, it is possible to change the value so that it best matches the requirements.

There is no use case, in form of real in code use, in this PR that validates the addition (tests do not count).
All extensions, that are currently not required, should be moved to separate PR; this is extension to structure that takes memory space and has not been proven, in code change, to be needed so far.
Inclusion of `erase_block` and `read_only` should be in separate PR, and discussed there; discussion on them distracts from the main purpose of the PR: is this the valid and ready replacement for the current functionality of flash map API.

Laczen commented:
> There is no use case, in form of real in code use, in this PR that validates the addition (tests do not count).
> All extensions, that are currently not required, should be moved to separate PR; this is extension to structure that takes memory space and has not been proven, in code change, to be needed so far.
> Inclusion of `erase_block` and `read_only` should be in separate PR, and discussed there; discussion on them distracts from the main purpose of the PR: is this the valid and ready replacement for the current functionality of flash map API.

@de-nordic, thanks for the comment. This PR proposes a new api to work with partitions, it is not meant ONLY as a direct replacement of what we have now. At the moment we do NOT have a good way to work with partitions (as illustrated by the comment of @jfischer-no during the dev-review meeting). The PR aims at providing a API to work with partitions. It does however still takes into consideration that it is also to be used by the flash_map API so that a transition from flash_area to flash_partition is not impossible.

As this is a PR it is open for improvements and I am always open to discussions about why a certain solution/addition was chosen.

Regarding the addition of read-only: it has been possible for a long time to make a partition read-only, it was however neglected by the flash_area API -> this is a correction, that should be in the initial PR.

Regarding the addition of a erase-block-size: this does not increase the size at all, it reduces the size as there no longer is a requirement to find out the sector size using the `get_sectors` routine. Suppose in the present solution we need to get the sector-size and there is a `get_sectors` routine, everytime we need the `get_sectors` routine there will be a pointer to this routine inside the code. The size of this one pointer is what is taken up by the `erase-block-size` in the proposed solution, the `get_sectors` routine is removed from flash and this is what is saved.

The question is "are heterogeneous sectors needed inside a partition ?", my proposal answer is "No" and use that as a means to reduce code. If the answer is, "Yes, we need that because then I can ...", a revision of the API would be needed (I do not consider "this is how we have it at the moment" a valid "Yes").

If this is just a question about naming, would it not be better to specify a "default-erase-block-size" instead of specifying "erase-block-size", no problem.

Regarding other discussion topics: "should the `flash_partition_check` be inside the API ?" My answer would be "No", this is not a property of a partition but of an image inside a partition. So IMO `flash_partition_check` should be removed.

de-nordic commented:
Btw, everyone seems to forget about this: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/fs/littlefs/boards/nrf52840dk_nrf52840.overlay
This is file system level configuration for FS done via DTS.

> This PR proposes a new api to work with partitions, it is not meant ONLY as a direct replacement of what we have now. 

Putting the rename aside, which may be ok move, it does exactly what the original API does: it limits operations on flash to predefined range. I am actually considering this as a valid candidate.

> At the moment we do NOT have a good way to work with partitions (as illustrated by the comment of @jfischer-no during the dev-review meeting). 
As far as I can tell, @jfischer-no concern was towards lack of declaration of "block" devices rather than partitions.

> The PR aims at providing a API to work with partitions. It does however still takes into consideration that it is also to be used by the flash_map API so that a transition from flash_area to flash_partition is not impossible.

Again it logically does the same thing as the flash map API, however you name it.

> Regarding the addition of read-only: it has been possible for a long time to make a partition read-only, it was however neglected by the flash_area API -> this is a correction, that should be in the initial PR.

There is no reason to add this: the only reason that this may be useful is to detect that user has failed to correctly select partition to write and destroys data, something that should be detected during tests.
Adding this, with all code paths within Zephyr  that will be checking the constant, will affect all code that access flash via flash_partition API even if 99.9% of users will never have the bug or will fix it at test phase; which is programmer's error. 

> Regarding the addition of a erase-block-size: this does not increase the size at all, it reduces the size as there no longer is a requirement to find out the sector size using the `get_sectors` routine. Suppose in the present solution we need to get the sector-size and there is a `get_sectors` routine, everytime we need the `get_sectors` routine there will be a pointer to this routine inside the code. The size of this one pointer is what is taken up by the `erase-block-size` in the proposed solution, the `get_sectors` routine is removed from flash and this is what is saved.
>
> The question is "are heterogeneous sectors needed inside a partition ?"

Depends of designation: if it is for streamed, sequential, data, or image of an application or data that is read only then
no, because "sectors"/"pages" are erased as data crosses from one to another; the `erase_block` is then useless;

In case of file systems: yes;  you can not reliably write efficient file system when your sectors change in size and if system does not use direct flash/mtd/magnetic device sector size, it will affect wear and/or speed.
Btw. the problem of non-uniform "sectors"/"pages" is nearly completely relevant to internal flash of SoC.

The FAT file system is an example of the problem: we define sector, for compatibility, to be 512 bytes, but on nrf flash it is written to 4096 byte pages, this means that each time one of sectors is to be written, there is possibility that the entire page needs to be erased and rewritten. Addition of the "erase_block" in API that should only limit flash access by offset and size is another level this can be broken or misconfigured.

I guess that the `erase_block`  would collect several pages as one bigger chunk: not useful for image/stream writes; bad for file systems.

The `flash_pages_info`, the flash_area_get_sectors` may need fix instead.

> 
> If this is just a question about naming, would it not be better to specify a "default-erase-block-size" instead of specifying "erase-block-size", no problem.

None. The flash partitioning, as the any other partition scheme, like MBR, GPT, Amiga RDB, BSD label, and so on, should never dictate sector size (effectively erase block), it may dictate partition start alignment, but not sector size - this is always taken from the hardware for software to decide what to do with the info.

> Regarding other discussion topics: "should the `flash_partition_check` be inside the API ?" My answer would be "No", this is not a property of a partition but of an image inside a partition. So IMO `flash_partition_check` should be removed.

This may be moved to separate unit or as library function. The removal of functions that are considered non needed is separate issue.

nvlsianpu commented:
The flash layout API was established for support any flash page layout. I see no point to left this behind.
I can see place for erase_block as value which can help to simplify some solutions not aimed to support homogenous layouts. 

I strongly believe that @jfischer-no concerns on API that might be used by mass storage device are layer much above subject of this API. Such an API will probably use this API amongst others to implement much sophisticated behaviors.

Laczen commented:
> Btw, everyone seems to forget about this: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/fs/littlefs/boards/nrf52840dk_nrf52840.overlay
> This is file system level configuration for FS done via DTS.

You mentioned it yesterday, and probably it will need some work, but I don't see any reason why the proposed PR would be incompatible with this. In fact this shows even the power of the proposed method:
a. We introduce a way to work with partitions that is good for zephyr.
b. We convert everything that is in zephyr to use this API (fcb, nvs, littlefs, ...)
c. We convert flash_map to use flash_partitions.
During the transition all conversions can be tested and introduced separately and if something is broken it will be limited.
 
> > At the moment we do NOT have a good way to work with partitions (as illustrated by the comment of @jfischer-no during the dev-review meeting).
> As far as I can tell, @jfischer-no concern was towards lack of declaration of "block" devices rather than partitions.

I don't think so, why is flash_disk, using the flash API rather than flash_map? Because flash_map is crippled by it's interface going through some struct that puts all partitions in an array and then gets a partition from this struct. Also look at the littlefs code about getting the block size, line 557 of little_fs.c "There is no flash_area API to get the maximum page size, so we have to make one", and what is really needed is a page size because the code will not work when a partition is created spanning heterogeneous sectors.
The same is valid for nvs, because of the flash_map interface it is only used in the initialisation and at the time I even resisted against that.

> 
> > The PR aims at providing a API to work with partitions. It does however still takes into consideration that it is also to be used by the flash_map API so that a transition from flash_area to flash_partition is not impossible.
> 
> Again it logically does the same thing as the flash map API, however you name it.

As said by naming it differently you can make a save and controlled transition.

> 
> > Regarding the addition of read-only: it has been possible for a long time to make a partition read-only, it was however neglected by the flash_area API -> this is a correction, that should be in the initial PR.
> 
> There is no reason to add this: the only reason that this may be useful is to detect that user has failed to correctly select partition to write and destroys data, something that should be detected during tests.
> Adding this, with all code paths within Zephyr that will be checking the constant, will affect all code that access flash via flash_partition API even if 99.9% of users will never have the bug or will fix it at test phase; which is programmer's error.

Let's turn this around, should I file a bug report because flash_map does not respect a "read-only" flag that I have set in the dts ? Once this bug report is present it would be OK to add a "read-only" flag?

> 
> > Regarding the addition of a erase-block-size: this does not increase the size at all, it reduces the size as there no longer is a requirement to find out the sector size using the `get_sectors` routine. Suppose in the present solution we need to get the sector-size and there is a `get_sectors` routine, everytime we need the `get_sectors` routine there will be a pointer to this routine inside the code. The size of this one pointer is what is taken up by the `erase-block-size` in the proposed solution, the `get_sectors` routine is removed from flash and this is what is saved.
> > The question is "are heterogeneous sectors needed inside a partition ?"
> 
> Depends of designation: if it is for streamed, sequential, data, or image of an application or data that is read only then
> no, because "sectors"/"pages" are erased as data crosses from one to another; the `erase_block` is then useless;
> 
> In case of file systems: yes; you can not reliably write efficient file system when your sectors change in size and if system does not use direct flash/mtd/magnetic device sector size, it will affect wear and/or speed.
> Btw. the problem of non-uniform "sectors"/"pages" is nearly completely relevant to internal flash of SoC.
> 
> The FAT file system is an example of the problem: we define sector, for compatibility, to be 512 bytes, but on nrf flash it is written to 4096 byte pages, this means that each time one of sectors is to be written, there is possibility that the entire page needs to be erased and rewritten. Addition of the "erase_block" in API that should only limit flash access by offset and size is another level this can be broken or misconfigured.
> 
> I guess that the `erase_block` would collect several pages as one bigger chunk: not useful for image/stream writes; bad for file systems.
> 
> The `flash_pages_info`, the flash_area_get_sectors` may need fix instead.
> 
> > If this is just a question about naming, would it not be better to specify a "default-erase-block-size" instead of specifying "erase-block-size", no problem.
> 
> None. The flash partitioning, as the any other partition scheme, like MBR, GPT, Amiga RDB, BSD label, and so on, should never dictate sector size (effectively erase block), it may dictate partition start alignment, but not sector size - this is always taken from the hardware for software to decide what to do with the info.
>

So there is no discussion that this would reduce code size rather then increase.

In the API there is no "dictation" of sector size or erase block size, it is a user configurable entry that can be used to help the user of the partition to get a well formulated block size or sector size. The default if not specified is the complete sector size. In most cases I would select is as the page size, but for nvs, fcb it makes sense to have fewer sectors so I would increase it if wanted, effectively using the erase-block-size to control the number of sectors needed. This would remove all code related to defining and checking the number of sectors, again a reduction in code size.

Introducing this one property gives many possible code reductions, and if there are users that would like more fine grained info about the real hardware limits, it is always possible.

From my knowledge of the zephyr system and flash usage this would reduce code size in 100% of the cases and in the other (using non-heterogeneous blocks !) the increase in code size is 1 word. 


  
> > Regarding other discussion topics: "should the `flash_partition_check` be inside the API ?" My answer would be "No", this is not a property of a partition but of an image inside a partition. So IMO `flash_partition_check` should be removed.
> 
> This may be moved to separate unit or as library function. The removal of functions that are considered non needed is separate issue.
It is not a separate issue when introducing a new API.

de-nordic commented:
> 
> 
> The flash layout API was established for support any flash page layout. I see no point to left this behind.
> I can see place for erase_block as value which can help to simplify some solutions not aimed to support homogenous layouts.

Actually the the flash_page info functions are only missing confirmation whether there is uniform page size across the device;
otherwise the get by offset and get by idx functions are all that is required.

Laczen commented:
> The flash layout API was established for support any flash page layout. I see no point to left this behind.
> I can see place for erase_block as value which can help to simplify some solutions not aimed to support homogenous layouts.
>
IMO the reason this was used in mynewt is that they were working with the complete flash and not with a partition.

> I strongly believe that @jfischer-no concerns on API that might be used by mass storage device are layer much above subject of this API. Such an API will probably use this API amongst others to implement much sophisticated behaviors.
I think the comment originated from the problem working with partitions and the direct usage of the flash API.

de-nordic commented:
> IMO the reason this was used in mynewt is that they were working with the complete flash and not with a partition.

If they have been using "complete flash" why do they need flash areas in the first place? and once more flash areas == flash partitions.

Laczen commented:
> > IMO the reason this was used in mynewt is that they were working with the complete flash and not with a partition.
> 
> If they have been using "complete flash" why do they need flash areas in the first place? and once more flash areas == flash partitions.

If they were defining areas as just an offset. Areas are a naming that was inherited by copying this from mynewt.

Laczen commented:
But while going through the code, I think there is one subsystem that allows the use of heterogeneous blocks: fcb, this however means that we should not use fcb for settings, as there is a possibility of losing data when we have been writing to a large sector and we have to copy all data in a smaller sector.

de-nordic commented:
> I don't think so, why is flash_disk, using the flash API rather than flash_map? Because flash_map is crippled by it's interface going through some struct that puts all partitions in an array and then gets a partition from this struct. 

Do you even read what I write? I am not talking about keeping the current flash map API way of obtaining the flash area object, or how the device has been obtain for each operation; that is why I have tried to resolve to compile time generated device pointers in the first time, and that is why I did my PR and that is why I consider this also to be candidate.

General idea of what you do is exactly the same as flash area, it is just a fence over some flash device that says where you can play. The execution is different.

> Also look at the littlefs code about getting the block size, line 557 of little_fs.c "There is no flash_area API to get the maximum page size, so we have to make one", and what is really needed is a page size because the code will not work when a partition is created spanning heterogeneous sectors.

It does not say: "we are missing constant with the value" but rather that there is no lib call that will report max page size in given range. So no It does not state we need `erase_block` set to some value, but rather that there is a need for  simpler way to find *actual* maximum value, within the area range, instead of doing this loop.

The above has also one advantage over just setting some value via DTS: switching partition to other device does not require altering the value of erase block, because driver will figure out.

We have `flash_get_page_info_by_offs` and `flash_get_page_info_by_idx` that are able to get page info, we could probably do `flash_partition_get_page_info_by_offs` , etc to mimic the flash_* API to do the same with the range of flash_partition, and simplify the access. But this is not an issue for the PR.

> and what is really needed is a page size because the code will not work when a partition is created spanning heterogeneous sectors.

So how does it fix the problem? It just hides it.

> Let's turn this around, should I file a bug report because flash_map does not respect a "read-only" flag that I have set in the dts ? Once this bug report is present it would be OK to add a "read-only" flag?

Maybe rather to remove the flag? The bug report you propose will spawn next one, to support it in `flash_partition_write` that will basically affect all write operations just for someone to figure out that the label in `FLASH_PARTITION_GET(get)` is incorrect, yet the execution path, within `flash_partition_write`, will remain, and will be basically code executed for no reason for 99.9% of users, for every write via API.

> So there is no discussion that this would reduce code size rather then increase.

I have not commented on size of code anywhere, and this is speculation that the PR is not proving, that is why, again, the extensions should be separate PR, with some change to exiting solution to show how to use the feature.

> In the API there is no "dictation" of sector size or erase block size, it is a user configurable entry that can be used to help the user of the partition to get a well formulated block size or sector size. (...)

It kinda is, because it has to be supported  internally within `flash_partition_erase` and probably `flash_partition_write`, otherwise what is the sense?
If the user has to make own calculations then the user may just pick any value anyway, and the `write_block` is no more useful than `#define`; in both cases user may just go for `flash_pages_info`.
 
> From my knowledge of the zephyr system and flash usage this would reduce code size in 100% of the cases and in the other (using non-heterogeneous blocks !) the increase in code size is 1 word.

From my knowledge this requires separate PR for discussion, so that we can focus on taking (or not taking) the flash_partition API as de-facto back-end for flash map API.

d3zd3z commented:
Regarding handling flash devices where the erase blocks vary in size. So far, everything I can find that uses MCUboot on these types of devices, always limits it to a set of sectors that have the same erase size. Although MCUboot has a bunch of code to try and support partitions that don't have constant sector sizes, I'm not sure how well this actually gets tested. The only configuration I see as needing to be supported is when the upgrade slot is an external device, in which case it can have a different erase size than the main device.

I would suggest we try to propose simplifying the use of variable erase block sizes within MCUboot, and if we can successfully do that, then having this new API just return a simple erase block size.

This could still be supported on devices with varying block sizes, as long each individual partition does not cross a boundary between sizes.  For example an ST device has 4 16K blocks, a 64K block, and 3 128K blocks. Typically, MCUboot resides in a 64k block, and the 3 128k blocks are used for images. The 16k blocks are used for persistent storage.

If anyone is aware of an MCUboot configuration where it tries to store code images in a partition where the block differs, please let us know.

Laczen commented:
To summarize the remarks on the PR for using flash partitions:

a. Does the flash partition API need to provide support for read-only partitions ?
b. Should the flash partition API always use the erase block size retrieved from a flash page API or should it be possible to specify a block size in the partition configuration (by no means limiting the usage to only this block size) ?
c. Should a flash partition API provide a image hash validation routine (the hash is not stored on the partition) ?

If there is extra info needed on the questions, do not hesitate.

d3zd3z commented:
> b. Should the flash partition API always use the erase block size retrieved from a flash page API or should it be possible to specify a block size in the partition configuration (by no means limiting the usage to only this block size) ?

It makes me uncomfortable when the flash API doesn't convey information about the true characteristics of the underlying flash device. Sometimes, this gets implemented writes that silently read/erase/write with the idea that they will help out application code that isn't expecting flash devices to be complicated. This can result in devastating data loss if power is lost between the erase and the data being rewritten. The loss situations of lying about erase blocks seems less significant, but there are still possible scenarios.

The conclusion I draw from this, is that if we do decide to drop the variable sector size mapping part of the API, then we are also asserting that partitions can never contains blocks of different sizes. For example, in the ST 4*16+64+3*128 device, there could be one or more partitions covering the 4 16K blocks, a partition covering the 64Kb block, and 1 or more covering the 3 128Kb blocks. But, a partition that includes say the 4 16K blocks and the 64K block (stating that there are just 2 64K blocks) should not be allowed.

As far as I can tell, all of our use of the ST device is this way, and I'm trying to investigate to see if there are any other configurations that people use. Given that MCUboot doesn't really test these varying-block-size partitions, I don't have high expectations that the code works, either.

Laczen commented:
@de-nordic, @nvlsianpu, @d3zd3z, I have been thinking about the variable block size problem, here is what I was thinking:

a. An optional parameter block-size can be specified for a partition.
b. When this block-size is specified `flash_partition_get_ebs()` returns this optional parameter.
c. When the block-size is not specified and FLASH_PAGE_LAYOUT is enabled `flash_partition_get_ebs()` returns the constant block size (if it is constant), if it is not constant it returns an error-code (`-EINVAL`).
d. A function `flash_partition_get_block_of(fs, off, &block)` that returns the block-offset (in the partition) and the block-size of the address `off` is pointing to. This function will be different based on the FLASH_PAGE_LAYOUT property.
e. When FLASH_PAGE_LAYOUT is not enabled and the optional parameter is not specified the partition is considered as one erasable block (so `..._get_ebs()` returns the partition size).

This way:
a. Variable block sizes are supported, and construction of "sectors" is also supported,
b. There is a way to detect a non-constant block size for a user that needs it,
c. There is a way to "short-circuit" the flash page layout when it is known that the device has a constant blocksize,
d. By default nothing changes if FLASH_PAGE_LAYOUT is selected.

nvlsianpu commented:
cc @erwango

nvlsianpu commented:
^^ That sound reasonably. 

Regard <b>b.</b> 

> When this block-size is specified flash_partition_get_ebs() returns this optional parameter.

Flash node in dts might also defined (native) block-size for the flash memory. Maybe it can be used if the partition node not defines it on its own.

Laczen commented:
@nvlsianpu, @de-nordic, I have updated to include support for non-uniform flash layout even in a partition.

To give an idea, the removal of flash_page support (by specifying the partition erase block size) saves about 400 bytes in rom.

de-nordic reviewed (changes requested):
The `erase_block` and the `read_only` require separate PR, because there are various things we do not agree on.
We need to focus, within this PR, to review the solution as the replacement for flash_area* within zephyr, as there is no need to let both exist internally;  with some agreement on need to change the name of flash_area* to something more appropriate, this may be the PR to got with.

I will review the rest of the code later, so far -1; I am just asking to have separate PR on extensions that can be added anytime but require some discussion and are currently distracting from quick adaptation of the API.

nvlsianpu commented:
@de-nordic @Laczen 

> I am just asking to have separate PR on extensions that can be added anytime but require some discussion and are currently distracting from quick adaptation of the API.

I think worth to divide PR to more subjects, this will help to get some of them faster.  @de-nordic  As I already understood you would leverage this API in the flash_area reworking efforts.

Laczen commented:
> The `erase_block` and the `read_only` require separate PR, because there are various things we do not agree on.
> We need to focus, within this PR, to review the solution as the replacement for flash_area* within zephyr, as there is no need to let both exist internally; with some agreement on need to change the name of flash_area* to something more appropriate, this may be the PR to got with.
> 
> I will review the rest of the code later, so far -1; I am just asking to have separate PR on extensions that can be added anytime but require some discussion and are currently distracting from quick adaptation of the API.

@de-nordic, Your reaction is rude and impolite.

From the start I have stated that this PR was NOT a replacement or a quick adaptation of flash_area. It is a proposal API to work with partitions.

The proposal allows the writing and reading and erasing of blocks in a partition and also provides some other functions that can be handy when working with partitions on flash.

It also guarantees that when using this API there will be no accidental overwrites of data stored in read-only partitions.

It also adds a **optional** property that allows a partition to specify a erase-block-size that is larger than(or equal to) the erase-block-size that is dictated by the hardware erase block size. This is usefull for users that could take advantage of it and allows significant reduction in image size. Yes this can be wrongly defined, but so can a hardware pin in the dts, when someone modifies a dts it is expected that there is basic knowledge on hardware limitations. **There is no intention (and there never was) to limit erases to this specified block size.** 

As examples where this optional property can be used:
a. To define a sectorsize (and thereby sectorcount) in nvs and fcb,
b. To define a blocksize (if possible) on a partition that has different hardware erase-block-sizes internally, but is used by a system that expects constant block sizes,
c. To define a erase-block-size for a block device that would store multiple blocks on one erase-block and where the hardware page size is no good fit. As an example of this a UBI like logical block to physical block translator that would store 512 byte blocks on a 1024 byte page size system.
d....

@nvlsianpu, thank you for trying to negotiate, but I can't see any good reason to remove both read-only and the optional erase-block-size from the API.

@de-nordic, remove yourself as a reviewer.

jfischer-no commented:
The way this PR is, it is kind of like flash_area_v2.

I do not think it is, or should be, a subsystem. From the API point of view it should not be much different than a (raw) flash API, or better, it should be a **mandatory** part of (raw) flash API, and use like all drivers the device API. So in the simplest case there is only one partition that occupies the whole raw flash area (the same regarding flash_area).  So just like that:
_________________________
|  (x) Application, FS, ...
|  ------------------------------------  |
| (2)  flashdisc, DFU, NVS  (always operates on a partition e.g. partition storage))          |
| (1)   raw flash API (read, erase, write API to a flash paritition) |
| nor-flash, nand-flash, xyz-flash (SoC/SPI/I2C raw flash device) |
_________________________

Laczen commented:
> The way this PR is, it is kind of like flash_area_v2.
> 
> I do not think it is, or should be, a subsystem. From the API point of view it should not be much different than a (raw) flash API, or better, it should be a **mandatory** part of (raw) flash API, and use like all drivers the device API. In the simplest case there is only one partition that occupies the whole raw flash area (the same regarding flash_area). So just like that:
> 
> (x) Application, FS, ...
> (2) flashdisc, DFU, NVS (always operates on a partition e.g. partition storage))
> (1) raw flash API (read, erase, write API to a flash paritition)
> nor-flash, nand-flash, xyz-flash (SoC/SPI/I2C raw flash device)

@jfischer-no, thank you for adding your comments. Is my understanding correct that you are proposing to add this under drivers/flash_partition ? This indeed seems logic as we are defining the partitions in the dts. Also this way zephyr users would find this interface quicker.

The flash_partition API should indeed be the way higher-up systems are working with flash.

I don't know if it's possible/wanted to deny applications/higher-up systems direct access to the raw flash device interface. Denying this raw access would definitely make testing of the raw flash (flash_partition) API a lot more difficult.

jfischer-no commented:
> @jfischer-no, thank you for adding your comments. Is my understanding correct that you are proposing to add this under drivers/flash_partition ? This indeed seems logic as we are defining the partitions in the dts. Also this way zephyr users would find this interface quicker.

Rather `drivers/flash/mtd_partition.c` or `drivers/flash/mtd_common.c` or `drivers/flash/flash_common.c`

> 
> The flash_partition API should indeed be the way higher-up systems are working with flash.
> 

No, I think it should be upper part of flash driver API and the only interface to higher layer.

> I don't know if it's possible/wanted to deny applications/higher-up systems direct access to the raw flash device interface. Denying this raw access would definitely make testing of the raw flash (flash_partition) API a lot more difficult.

It is not denied, it still operates on raw flash but with advantages you mentioned above:
>The proposal allows the writing and reading and erasing of blocks in a partition and also provides some other functions that can be handy when working with partitions on flash.
It also guarantees that when using this API there will be no accidental overwrites of data stored in read-only partitions.

Laczen commented:
@jfischer-no, this seems like a good idea, I would go for mtd_partition.c. For the full fledge mtd support like in linux my knowledge of flash devices is to limited.

Should we go for a `mtd_write(struct *mtd_info, off, data, len)` approach or keep something like `mtd_partition_write(struct *mtd_partition, off, data,len)` ?

Laczen commented:
@jfischer-no, I have updated the proposal

I am considering defining `mtd_write(...)` routines instead of `mtd_part_write(...)` and renaming the `mtd_part` structure to `mtd_info`.

Also routine `mtd_create_partition()` could be added (it is just a struct definition).

nvlsianpu commented:
>
@nvlsianpu, thank you for trying to negotiate, but I can't see any good reason to remove both read-only and the optional erase-block-size from the API.

I see,
@Laczen Consent helps to have work done. Actually I was really thinking that these feature were nice and might be added in separate discussion (PRs). Point is that when we have to long discussion it is hard to get to the point of PR.
I bet that if it is easy to prove that  feature is useful, then it will be accepted anyway. From the other hand iterative development allow to divide effort into easier manageable chunks.

> @de-nordic, remove yourself as a reviewer.

That you both are not agreed on technical issue doesn't mean that anyone can exclude someone from discussion. Domink has spent much time on this and related subject, missing his review mean waste of his efforts and time.

nvlsianpu commented on drivers/flash/CMakeLists.txt:
```diff
@@ -21,6 +21,7 @@ zephyr_library_sources_ifdef(CONFIG_FLASH_STM32_QSPI flash_stm32_qspi.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_MX25UM51345G flash_mcux_flexspi_mx25um51345g.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_NOR flash_mcux_flexspi_nor.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_ESP32 flash_esp32.c)
+zephyr_library_sources_ifdef(CONFIG_MTD_PARTITIONING mtd_part.c)
```
I start thinking that this is more generic kind of driver and should be moved out of flash directory. It will be layer which can be gate to any memory, not only flash.

Laczen replied:
In linux I think the flash drivers are found under mtd instead of flash, under mtd there is a mtd_core that does something similar as what we need. Some of my inspiration comes from there. The mtd (devices) require a erase routine, on other devices this is not required.

nvlsianpu commented on tests/drivers/mtd_partition/native_posix.overlay:
```diff
@@ -0,0 +1,45 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				bootsub_partition: partition@4000 {
```
oh, a sub-partition. Why it is needed?

Laczen replied:
On linux there is no limitation that there can only be partitions and not subpartitions. Introducing subpartitions (this has always been possible, but there was no support to use them) will make a lot of things easier.

e.g. For image generation: It is possible to specify image0 as header + real_image0, generate the image for real_image0, upload to slot image0. When other bootloaders use a different scheme -> no problem. Very versatile.

Laczen commented:
@jfischer-no, updated the PR to allow working with mtd masters as well as partitions.
